src/fetch_blocks.rs

```diff
@@ -0,0 +1,395 @@
+            send: self.send.clone(),
+            connection: self.recv.try_next().ok().and_then(|a| a),
+            dirty: false,
+        }
+    }
+}
+impl std::fmt::Debug for Peer {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_struct("Peer").field("addr", &self.addr).finish()
+    }
+}
+
+pub struct PeerHandle {
+    addr: Arc<Address>,
+    send: mpsc::Sender<BitcoinPeerConnection>,
+    connection: Option<BitcoinPeerConnection>,
+    dirty: bool,
+}
+impl PeerHandle {
+    pub async fn with_connection<
+        T,
+        F: FnOnce(&mut BitcoinPeerConnection) -> Fut,
+        Fut: std::future::Future<Output = Result<T, Error>>,
+    >(
+        &mut self,
+        env: Arc<Env>,
+        f: F,
+    ) -> Result<T, Error> {
+        let conn = if let Some(ref mut connection) = self.connection {
+            connection
+        } else {
+            self.connection =
+                Some(BitcoinPeerConnection::connect(env, (&*self.addr).clone()).await?);
+            self.connection.as_mut().unwrap()
+        };
+        self.dirty = true;
+        f(conn).await
+    }
+
+    pub async fn with_connection_blocking<
+        T: Send + 'static,
+        F: FnOnce(&mut BitcoinPeerConnection) -> Result<T, Error> + Send + 'static,
+    >(
+        &mut self,
+        env: Arc<Env>,
+        f: F,
+    ) -> Result<T, Error> {
+        let mut conn = if let Some(connection) = self.connection.take() {
+            connection
+        } else {
+            BitcoinPeerConnection::connect(env, (&*self.addr).clone()).await?
+        };
+        self.dirty = true;
+        let (conn, res) = tokio::task::spawn_blocking(move || {
+            let res = f(&mut conn);
+            (conn, res)
+        })
+        .await?;
+        self.connection = Some(conn);
+        Ok(res?)
+    }
+
+    pub fn mark_clean(&mut self) {
+        self.dirty = false;
+    }
+}
+impl<'a> Drop for PeerHandle {
+    fn drop(&mut self) {
+        if !self.dirty {
+            if let Some(connection) = self.connection.take() {
+                self.send.try_send(connection).unwrap_or_default()
+            }
+        }
+    }
+}
+
+async fn fetch_block_from_self(env: &Env, hash: BlockHash) -> Result<Option<Block>, RpcError> {
+    match env
+        .rpc_client
+        .call(&RpcRequest {
+            id: None,
+            method: GetBlock,
+            params: GetBlockParams(hash, Some(0)),
+        })
+        .await?
+        .into_result()
+    {
+        Ok(b) => Ok(Some(
+            Block::consensus_decode(&mut std::io::Cursor::new(
+                b.as_left()
+                    .ok_or_else(|| anyhow::anyhow!("unexpected response for getblock"))?
+                    .as_ref(),
+            ))
+            .map_err(Error::from)?,
+        )),
+        Err(e) if e.code == MISC_ERROR_CODE && e.message == PRUNE_ERROR_MESSAGE => Ok(None),
+        Err(e) => Err(e),
+    }
+}
+
+async fn fetch_block_from_peer<'a>(
```

Comment:
the idea was that if an error (or panic) occurred before the connection was used, it would still be reusable, but in this scenario, we don't really use that, so this pattern isn't really useful.

Reply:
OK, I will look at it after merging and try to simplify it. Seems better than adding it into this already huge PR.